lib/routes/shared/security_pin/prompt_pin_code.dart

```diff
@@ -0,0 +1,298 @@
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+
+  Padding _buildBreezLogo() {
+    return Padding(
+      child: _breezLogo,
+      padding: EdgeInsets.only(top: widget.dismissible ? kToolbarHeight : 96.0, bottom: 96.0),
+    );
+  }
+
+  Container _buildPinCircles() {
+    return Container(
+      margin: const EdgeInsets.only(
+        top: 16,
+        left: 64,
+        right: 64,
+      ),
+      height: 48,
+      child: Row(
+        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+        children: _buildCircles(),
+      ),
+    );
+  }
+
+  Padding _buildErrorMessage() {
+    return _errorMessage.isNotEmpty
+        ? Padding(
+            padding: EdgeInsets.only(bottom: 64.0),
+            child: Text(
+              _errorMessage,
+              style: theme.errorStyle,
+            ),
+          )
+        : Padding(
+            padding: EdgeInsets.only(bottom: 64.0),
+            child: Text(
+              "",
+              style: theme.errorStyle,
+            ),
+          );
+  }
+
+  List<Widget> _buildCircles() {
+    var list = <Widget>[];
+    for (int i = 0; i < _passwordLength; i++) {
+      list.add(AnimatedContainer(
+        duration: Duration(milliseconds: 200),
+        padding: EdgeInsets.only(top: 24.0),
+        margin: EdgeInsets.only(bottom: 0),
+        width: _enteredPinCode.length == _passwordLength ? 28 : 24,
+        height: _enteredPinCode.length == _passwordLength ? 28 : 24,
+        decoration: BoxDecoration(
+            color: i < _enteredPinCode.length ? Colors.white : Colors.transparent,
+            shape: BoxShape.circle,
+            border: Border.all(color: Colors.white, width: 2.0)),
+      ));
+    }
+    return list;
+  }
+
+  Widget _numPad() {
+    return new GridView.count(
+        crossAxisCount: 3,
+        childAspectRatio: 5 / 3,
+        padding: EdgeInsets.zero,
+        children: List<int>.generate(9, (i) => i).map((index) => _numberButton((index + 1).toString())).followedBy([
+          Container(
+            child: new IconButton(
+              onPressed: () => _setPinCodeInput(""),
+              icon: Icon(
+                Icons.delete_forever,
+                color: Colors.white,
+              ),
+            ),
+          ),
+          _numberButton("0"),
+          Container(
+            child: new IconButton(
+              onPressed: () => _setPinCodeInput(_enteredPinCode.substring(0, _enteredPinCode.length - 1)),
+              icon: Icon(
+                Icons.backspace,
+                color: Colors.white,
+              ),
+            ),
+          ),
+        ]).toList());
+  }
+
+  Container _numberButton(String number) {
+    return Container(
+      child: new FlatButton(
+        onPressed: () => _onNumButtonPressed(number),
+        child: new Text(number, textAlign: TextAlign.center, style: theme.numPadNumberStyle),
+      ),
+    );
+  }
+
+  _onNumButtonPressed(String numberText) {
```

Comment:
This method code is too complicated, it took me some time to understand that it actually handles 3 use cases:
1. Enter PIN to unlock.
2. Change existing PIN.
3. Set new PIN.

And that the code differentiate the cases by these variables: `setPassword`, `changePassword`
I think we can simplify:
1. use cases 2,3 are the same. The difference is that the old PIN is empty in case of new pin so no need to validate against it before changing.
2. We can use one flag `changePIN` which when true it refers to "changing existing pin" case and when false it refers to "Enter to unlock" case.

We will need to iterate again after this simplification. It needs to be very clear.

Reply:
One more thing that will simplify a lot is to separate the UI of the pinCode from the use cases that are handled in the stateful widget. 
I suggest to create a StatelessWidget and move all the stateless UI there (numpad, circles, errorMessage, etc...).
We can call it `PinCodeWidget` and can put it in the general widgets directory.
Then we can leave the validation and controller code here in the stateful component and see if it looks simple enough or we need to separate further.